fdbserver/VersionedBTree.actor.cpp

```diff
@@ -2871,50 +2942,60 @@ class DWALPager final : public IPager2 {
 	                                                           bool header) {
 		ASSERT(!self->memoryOnly);
 
-		// if (g_network->getCurrentTask() > TaskPriority::DiskRead) {
-		// 	wait(delay(0, TaskPriority::DiskRead));
-		// }
-
 		state Reference<ArenaPage> page =
-		    header ? Reference<ArenaPage>(new ArenaPage(smallestPhysicalBlock, smallestPhysicalBlock))
-		           : self->newPageBuffer();
-		debug_printf("DWALPager(%s) op=readPhysicalStart %s ptr=%p\n",
+		    header ? ArenaPage::construct(smallestPhysicalBlock, smallestPhysicalBlock) : self->newPageBuffer();
+		debug_printf("DWALPager(%s) op=readPhysicalStart %s ptr=%p header=%d\n",
 		             self->filename.c_str(),
 		             toString(pageID).c_str(),
-		             page->begin());
-
-		state PriorityMultiLock::Lock lock = wait(self->ioLock.lock(std::min(priority, ioMaxPriority)));
-		++g_redwoodMetrics.metric.pagerDiskRead;
-
-		// TODO:  Could a dispatched read try to write to page after it has been destroyed if this actor is cancelled?
-		int blockSize = header ? smallestPhysicalBlock : self->physicalPageSize;
-		int readBytes = wait(self->pageFile->read(page->mutate(), blockSize, (int64_t)pageID * blockSize));
-		debug_printf("DWALPager(%s) op=readPhysicalComplete %s ptr=%p bytes=%d\n",
+		             page->rawData(),
+		             header);
+
+		int readBytes = wait(readPhysicalBlock(self,
+		                                       page->rawData(),
+		                                       page->rawSize(),
+		                                       (int64_t)pageID * page->rawSize(),
+		                                       std::min(priority, ioMaxPriority)));
+		debug_printf("DWALPager(%s) op=readPhysicalDiskReadComplete %s ptr=%p bytes=%d\n",
 		             self->filename.c_str(),
 		             toString(pageID).c_str(),
-		             page->begin(),
+		             page->rawData(),
 		             readBytes);
 
-		// Header reads are checked explicitly during recovery
-		if (!header) {
-			if (!page->verifyChecksum(pageID)) {
-				debug_printf(
-				    "DWALPager(%s) checksum failed for %s\n", self->filename.c_str(), toString(pageID).c_str());
-				Error e = checksum_failed();
-				if (g_network->isSimulated() && g_simulator.checkInjectedCorruption())
-					e = e.asInjectedFault();
-				TraceEvent(SevError, "RedwoodChecksumFailed")
-				    .detail("Filename", self->filename.c_str())
-				    .detail("PageID", pageID)
-				    .detail("PageSize", self->physicalPageSize)
-				    .detail("Offset", pageID * self->physicalPageSize)
-				    .detail("CalculatedChecksum", page->calculateChecksum(pageID))
-				    .detail("ChecksumInPage", page->getChecksum())
-				    .error(e);
-				ASSERT(false);
-				throw e;
+		try {
+			page->postRead1(pageID);
```

Comment:
a nit: we should rename "postRead1/2" to more meaningful names; code isn't very readable in the current state, IMHO.

Reply:
Good idea.

Reply:
Josh had the same feedback, with name suggestions, so I changed it.